File: app/src/main/java/com/example/bloodbank/Model/DonorAchievement.java
package com.example.bloodbank.Model;

import java.util.HashMap;
import java.util.Map;

public class DonorAchievement {
    private String donorId;
    private int totalPoints;
    private Map<String, Boolean> badges;
    private int donationStreak;
    private long lastDonationDate;
    private int rank;
    private String currentTitle;
    private Map<String, Long> achievementDates;

    public DonorAchievement() {
        // Required empty constructor for Firebase
        this.badges = new HashMap<>();
        this.achievementDates = new HashMap<>();
        this.totalPoints = 0;
        this.donationStreak = 0;
        this.rank = 0;
        this.currentTitle = "Novice Donor";
    }

    public DonorAchievement(String donorId) {
        this();
        this.donorId = donorId;
    }

    // Points system
    public void addPoints(int points) {
        this.totalPoints += points;
        updateTitle();
    }

    private void updateTitle() {
        if (totalPoints >= 1000) {
            currentTitle = "Legendary Lifesaver";
        } else if (totalPoints >= 500) {
            currentTitle = "Elite Donor";
        } else if (totalPoints >= 250) {
            currentTitle = "Regular Hero";
        } else if (totalPoints >= 100) {
            currentTitle = "Blood Champion";
        } else if (totalPoints >= 50) {
            currentTitle = "Active Donor";
        }
    }

    // Badge system
    public void unlockBadge(String badgeName) {
        badges.put(badgeName, true);
        achievementDates.put(badgeName, System.currentTimeMillis());
    }

    public boolean hasBadge(String badgeName) {
        return badges.containsKey(badgeName) && badges.get(badgeName);
    }

    // Streak system
    public void updateDonationStreak(long donationDate) {
        // If it's been more than 4 months since last donation, reset streak
        if (lastDonationDate > 0 &&
                (donationDate - lastDonationDate) > (120L * 24 * 60 * 60 * 1000)) {
            donationStreak = 1;
        } else {
            donationStreak++;
        }
        lastDonationDate = donationDate;
    }

    // Getters and Setters
    public String getDonorId() {
        return donorId;
    }

    public void setDonorId(String donorId) {
        this.donorId = donorId;
    }

    public int getTotalPoints() {
        return totalPoints;
    }

    public void setTotalPoints(int totalPoints) {
        this.totalPoints = totalPoints;
        updateTitle();
    }

    public Map<String, Boolean> getBadges() {
        return badges;
    }

    public void setBadges(Map<String, Boolean> badges) {
        this.badges = badges;
    }

    public int getDonationStreak() {
        return donationStreak;
    }

    public void setDonationStreak(int donationStreak) {
        this.donationStreak = donationStreak;
    }

    public long getLastDonationDate() {
        return lastDonationDate;
    }

    public void setLastDonationDate(long lastDonationDate) {
        this.lastDonationDate = lastDonationDate;
    }

    public int getRank() {
        return rank;
    }

    public void setRank(int rank) {
        this.rank = rank;
    }

    public String getCurrentTitle() {
        return currentTitle;
    }

    public void setCurrentTitle(String currentTitle) {
        this.currentTitle = currentTitle;
    }

    public Map<String, Long> getAchievementDates() {
        return achievementDates;
    }

    public void setAchievementDates(Map<String, Long> achievementDates) {
        this.achievementDates = achievementDates;
    }
}

File: app/src/main/java/com/example/bloodbank/Model/DonorAchievement.kt
package com.example.bloodbank.Model

import androidx.annotation.Keep

/**
 * Donor achievement data model for gamification system.
 * Tracks donor points, badges, streaks, and titles.
 */
@Keep
data class DonorAchievement(
    var donorId: String = "",
    var totalPoints: Int = 0,
    var badges: MutableMap<String, Boolean> = mutableMapOf(),
    var donationStreak: Int = 0,
    var lastDonationDate: Long = 0L,
    var rank: Int = 0,
    var currentTitle: String = "Novice Donor",
    var achievementDates: MutableMap<String, Long> = mutableMapOf()
) {
    // Required empty constructor for Firebase
    constructor() : this("", 0, mutableMapOf(), 0, 0L, 0, "Novice Donor", mutableMapOf())
    
    // Secondary constructor for creating with just donorId
    constructor(donorId: String) : this(
        donorId = donorId,
        totalPoints = 0,
        badges = mutableMapOf(),
        donationStreak = 0,
        lastDonationDate = 0L,
        rank = 0,
        currentTitle = "Novice Donor",
        achievementDates = mutableMapOf()
    )
    
    /**
     * Add points to the donor's total and update their title.
     */
    fun addPoints(points: Int) {
        totalPoints += points
        updateTitle()
    }
    
    /**
     * Update the donor's title based on their total points.
     */
    private fun updateTitle() {
        currentTitle = when {
            totalPoints >= 1000 -> "Legendary Lifesaver"
            totalPoints >= 500 -> "Elite Donor"
            totalPoints >= 250 -> "Regular Hero"
            totalPoints >= 100 -> "Blood Champion"
            totalPoints >= 50 -> "Active Donor"
            else -> "Novice Donor"
        }
    }
    
    /**
     * Unlock a badge for the donor.
     */
    fun unlockBadge(badgeName: String) {
        badges[badgeName] = true
        achievementDates[badgeName] = System.currentTimeMillis()
    }
    
    /**
     * Check if the donor has a specific badge.
     */
    fun hasBadge(badgeName: String): Boolean {
        return badges[badgeName] == true
    }
    
    /**
     * Update the donation streak based on the donation date.
     * Resets if more than 4 months have passed since last donation.
     */
    fun updateDonationStreak(donationDate: Long) {
        // If it's been more than 4 months since last donation, reset streak
        val fourMonthsInMillis = 120L * 24 * 60 * 60 * 1000
        donationStreak = if (lastDonationDate > 0 && (donationDate - lastDonationDate) > fourMonthsInMillis) {
            1
        } else {
            donationStreak + 1
        }
        lastDonationDate = donationDate
    }
    
    /**
     * Custom setter for totalPoints that updates the title.
     */
    fun setTotalPointsWithUpdate(points: Int) {
        totalPoints = points
        updateTitle()
    }
}
